File: mobile/ios/OpenU60/Features/RouterSettings/SIM/SIMViewModel.swift
import SwiftUI

enum PinSheetAction {
    case verify        // unlock PIN-locked SIM
    case enableLock    // enable PIN lock on SIM
    case disableLock   // disable PIN lock on SIM
}

@Observable
@MainActor
final class SIMViewModel {
    var simInfo: SIMInfo = .empty
    var lockInfo: SIMLockInfo = .empty
    var isLoading: Bool = false
    var message: String?
    var messageIsError: Bool = false

    // Sheet state
    var showChangePinSheet: Bool = false
    var showEnterPinSheet: Bool = false
    var showEnterPukSheet: Bool = false
    var showUnlockSheet: Bool = false
    var pinSheetAction: PinSheetAction = .verify

    // Form fields
    var pinInput: String = ""
    var oldPinInput: String = ""
    var newPinInput: String = ""
    var pukInput: String = ""
    var nckInput: String = ""

    private let client: AgentClient
    private let authManager: AuthManager

    init(client: AgentClient, authManager: AuthManager) {
        self.client = client
        self.authManager = authManager
    }

    var isPinEnabled: Bool {
        simInfo.pinStatus == "1"
    }

    var isPinLocked: Bool {
        let sim = simInfo.simStatus.lowercased()
        let modem = simInfo.modemMainState.lowercased()
        return sim == "wait pin" || modem == "modem_waitpin"
    }

    var isPukLocked: Bool {
        let sim = simInfo.simStatus.lowercased()
        let modem = simInfo.modemMainState.lowercased()
        return sim == "wait puk" || modem == "modem_waitpuk"
    }

    func submitPin() async {
        switch pinSheetAction {
        case .verify:
            await verifyPin()
        case .enableLock:
            await changePinMode(enable: true)
            showEnterPinSheet = false
        case .disableLock:
            await changePinMode(enable: false)
            showEnterPinSheet = false
        }
    }

    func refresh() async {
        isLoading = true
        message = nil

        async let simTask = fetchSIMInfo()
        async let lockTask = fetchSIMLock()

        let (sim, lock) = await (simTask, lockTask)
        if let sim { simInfo = sim }
        if let lock { lockInfo = lock }

        isLoading = false
    }

    func changePinMode(enable: Bool) async {
        isLoading = true

        do {
            let _ = try await client.postJSON("/api/sim/pin/mode", body: [
                    "pin_num_m": pinInput,
                    "pin_mode": enable ? 1 : 0,
                    "pin_encode_flag": "0"
                ])
            pinInput = ""
            showMessage(enable ? "PIN 锁已启用" : "PIN 锁已禁用", isError: false)
            await refresh()
        } catch {
            showMessage("失败：\(error.localizedDescription)", isError: true)
        }

        isLoading = false
    }

    func changePin() async {
        guard oldPinInput.count >= 4, newPinInput.count >= 4 else {
            showMessage("PIN 至少需要 4 位数字", isError: true)
            return
        }

        isLoading = true

        do {
            let _ = try await client.postJSON("/api/sim/pin/change", body: [
                    "pin_num": oldPinInput,
                    "new_pin_num": newPinInput,
                    "pin_encode_flag": "0"
                ])
            oldPinInput = ""
            newPinInput = ""
            showChangePinSheet = false
            showMessage("PIN 修改成功", isError: false)
        } catch {
            showMessage("失败：\(error.localizedDescription)", isError: true)
        }

        isLoading = false
    }

    func verifyPin() async {
        guard pinInput.count >= 4 else {
            showMessage("PIN 至少需要 4 位数字", isError: true)
            return
        }

        isLoading = true

        do {
            let _ = try await client.postJSON("/api/sim/pin/verify", body: [
                    "pin_num": pinInput,
                    "puk_num": "",
                    "pin_encode_flag": "0"
                ])
            pinInput = ""
            showEnterPinSheet = false
            showMessage("PIN 验证成功", isError: false)
            await refresh()
        } catch {
            showMessage("失败：\(error.localizedDescription)", isError: true)
        }

        isLoading = false
    }

    func verifyPuk() async {
        guard pukInput.count >= 8 else {
            showMessage("PUK 至少需要 8 位数字", isError: true)
            return
        }
        guard newPinInput.count >= 4 else {
            showMessage("新 PIN 至少需要 4 位数字", isError: true)
            return
        }

        isLoading = true

        do {
            let _ = try await client.postJSON("/api/sim/pin/verify", body: [
                    "pin_num": newPinInput,
                    "puk_num": pukInput,
                    "pin_encode_flag": "0"
                ])
            pukInput = ""
            newPinInput = ""
            showEnterPukSheet = false
            showMessage("PUK 验证成功，新 PIN 已设置", isError: false)
            await refresh()
        } catch {
            showMessage("失败：\(error.localizedDescription)", isError: true)
        }

        isLoading = false
    }

    func unlockSIM() async {
        guard !nckInput.isEmpty else {
            showMessage("需要输入解锁码", isError: true)
            return
        }

        isLoading = true

        do {
            let _ = try await client.postJSON("/api/sim/unlock", body: ["nck": nckInput])
            nckInput = ""
            showUnlockSheet = false
            showMessage("SIM 解锁成功", isError: false)
            await refresh()
        } catch {
            showMessage("失败：\(error.localizedDescription)", isError: true)
        }

        isLoading = false
    }

    // MARK: - Private

    private func fetchSIMInfo() async -> SIMInfo? {
        do {
            let data = try await client.getJSON("/api/sim/info")
            return SIMParser.parseSIMInfo(data)
        } catch {
            showMessage("加载 SIM 信息失败：\(error.localizedDescription)", isError: true)
            return nil
        }
    }

    private func fetchSIMLock() async -> SIMLockInfo? {
        do {
            let data = try await client.getJSON("/api/sim/lock-trials")
            return SIMParser.parseSIMLock(data)
        } catch {
            return nil
        }
    }

    private func showMessage(_ text: String, isError: Bool) {
        message = text
        messageIsError = isError
    }
}

File: mobile/android/OpenU60/app/src/main/java/com/openu60/feature/router/sim/SIMViewModel.kt
package com.openu60.feature.router.sim

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.openu60.core.model.SIMInfo
import com.openu60.core.model.SIMLockInfo
import com.openu60.core.model.SIMParser
import com.openu60.core.network.AgentClient
import com.openu60.core.network.AgentError
import com.openu60.core.network.AuthManager
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.async
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class SIMState(
    val simInfo: SIMInfo = SIMInfo.empty,
    val lockInfo: SIMLockInfo = SIMLockInfo.empty,
    val isLoading: Boolean = false,
    val message: String? = null,
    val messageIsError: Boolean = false,
)

@HiltViewModel
class SIMViewModel @Inject constructor(
    private val agentClient: AgentClient,
    private val authManager: AuthManager,
) : ViewModel() {

    private val _state = MutableStateFlow(SIMState())
    val state: StateFlow<SIMState> = _state.asStateFlow()

    fun refresh() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                val infoDeferred = async { agentClient.getJSON("/api/sim/info") }
                val lockDeferred = async { agentClient.getJSON("/api/sim/lock-trials") }
                val simInfo = SIMParser.parseSIMInfo(infoDeferred.await())
                val lockInfo = SIMParser.parseSIMLock(lockDeferred.await())
                _state.value = _state.value.copy(simInfo = simInfo, lockInfo = lockInfo, isLoading = false)
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) refresh() else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun verifyPIN(pin: String) {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                agentClient.postJSON("/api/sim/pin/verify", mapOf(
                    "pin_num" to pin,
                    "puk_num" to "",
                    "pin_encode_flag" to "0",
                ))
                _state.value = _state.value.copy(isLoading = false, message = "PIN 验证成功", messageIsError = false)
                refresh()
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) verifyPIN(pin) else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun changePIN(oldPin: String, newPin: String) {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                agentClient.postJSON("/api/sim/pin/change", mapOf(
                    "pin_num" to oldPin,
                    "new_pin_num" to newPin,
                    "pin_encode_flag" to "0",
                ))
                _state.value = _state.value.copy(isLoading = false, message = "PIN 已修改", messageIsError = false)
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) changePIN(oldPin, newPin) else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun togglePINMode(enabled: Boolean, pin: String) {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                agentClient.postJSON("/api/sim/pin/mode", mapOf(
                    "pin_num_m" to pin,
                    "pin_mode" to if (enabled) 1 else 0,
                    "pin_encode_flag" to "0",
                ))
                _state.value = _state.value.copy(
                    isLoading = false,
                    message = if (enabled) "PIN 锁已启用" else "PIN 锁已禁用",
                    messageIsError = false,
                )
                refresh()
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) togglePINMode(enabled, pin) else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun verifyPUK(puk: String, newPin: String) {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                agentClient.postJSON("/api/sim/pin/verify", mapOf(
                    "pin_num" to newPin,
                    "puk_num" to puk,
                    "pin_encode_flag" to "0",
                ))
                _state.value = _state.value.copy(isLoading = false, message = "PUK 验证成功，PIN 已重置", messageIsError = false)
                refresh()
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) verifyPUK(puk, newPin) else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun unlockNCK(nck: String) {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                agentClient.postJSON("/api/sim/unlock", mapOf("nck" to nck))
                _state.value = _state.value.copy(isLoading = false, message = "SIM 已解锁", messageIsError = false)
                refresh()
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) unlockNCK(nck) else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    private fun setError(msg: String?) {
        _state.value = _state.value.copy(isLoading = false, message = msg ?: "未知错误", messageIsError = true)
    }
}
